samza-kafka/src/main/scala/org/apache/samza/checkpoint/kafka/KafkaCheckpointManager.scala

```diff
@@ -286,15 +259,11 @@ class KafkaCheckpointManager(
   }
 
   def stop = {
-    if (producer != null) {
-      producer.close
+    if (systemProducer != null) {
+      systemProducer.stop
```

Comment:
Should synchronize all accesses to `systemProducer` for memory visibility.
Consider two threads: T1, T2. T1 recycles the producer reference, which is not visible to T2. T2 calls `stop` on the previous producer instance causing a leak?

Reply:
yep. Done.